pkg/cluster/fixetcd.go

```diff
@@ -0,0 +1,517 @@
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      jobName,
+			Namespace: nameSpace,
+			Labels:    map[string]string{"app": jobName},
+		},
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      jobName,
+					Namespace: nameSpace,
+					Labels:    map[string]string{"app": jobName},
+				},
+				Spec: corev1.PodSpec{
+					RestartPolicy: corev1.RestartPolicyOnFailure,
+					NodeName:      node,
+					Containers: []corev1.Container{
+						{
+							Name:  "backup",
+							Image: image,
+							Command: []string{
+								"chroot",
+								"/host",
+								"/bin/bash",
+								"-c",
+								shellScriptFixEtcd,
+							},
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									Name:      "host",
+									MountPath: "/host",
+									ReadOnly:  false,
+								},
+							},
+							SecurityContext: &corev1.SecurityContext{
+								Capabilities: &corev1.Capabilities{
+									Add: []corev1.Capability{"SYS_CHROOT"},
+								},
+								Privileged: to.BoolPtr(true),
+							},
+							Env: []corev1.EnvVar{
+								{
+									Name:  "BACKUP",
+									Value: "true",
+								},
+							},
+						},
+					},
+					Volumes: []corev1.Volume{
+						{
+							Name: "host",
+							VolumeSource: corev1.VolumeSource{
+								HostPath: &corev1.HostPathVolumeSource{
+									Path: "/",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+
+	w, err := m.kubernetescli.BatchV1().Jobs(b.Namespace).Watch(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	time.Sleep(30 * time.Second)
+
+	wc := w.ResultChan()
+	<-wc
+
+	m.kubernetescli.BatchV1().Jobs(b.Namespace).Delete(ctx, jobName, metav1.DeleteOptions{})
+	w.Stop()
+
+	return nil
+}
+
+func (m *manager) checkEtcds(ctx context.Context, e *operatorv1.Etcd) bool {
+	for _, c := range e.Status.Conditions {
+		if c.Type == etcdEndpointsDegraded && c.Status == operatorv1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+func (m *manager) comparePodEnvToIP(pods *corev1.PodList) *degradedEtcd {
+	de := &degradedEtcd{}
+	for _, p := range pods.Items {
+		etcdIP := m.ipFromEnv(p.Spec.Containers, p.Name)
+		for _, ip := range p.Status.PodIPs {
+			if ip.IP != etcdIP && etcdIP != "" {
+				m.log.Infof("Found non matching IPs for etcd Pod %s: %s!=%s", p.Name, ip.IP, etcdIP)
+				return &degradedEtcd{
+					Node:  strings.ReplaceAll(p.Name, "etcd-", ""),
+					Pod:   p.Name,
+					NewIP: ip.IP,
+					OldIP: etcdIP,
+				}
+			}
+		}
+	}
+	return de
+}
+
+func (m *manager) ipFromEnv(containers []corev1.Container, podName string) string {
+	for _, c := range containers {
+		if c.Name == "etcd" {
+			for _, e := range c.Env {
+				envName := strings.ReplaceAll(podName, "-", "_")
+				envName = strings.ReplaceAll(envName, "etcd_", "NODE_")
```

Comment:
It's used within `comparePodEnvToIp`, I made an additional function to keep the former easier to understand/read.

Reply:
Because the pod's environment variable names are almost the same as `podName`. The differences being the name contains underscores instead of dashes and prefixed with `NODE_` rather than `etcd_`.